### ISSUE TITLE
Issue with Pose estimation video compared to generated Tracklets

### ISSUE THREAD

**Issue by CalvinLloyd117** [DeepLabCut/DeepLabCut#937]
Thanks for opening this issue, and thanks for using DeepLabCut (we hope you are enjoying it ;). Please fill out the template completely, including the full "traceback" and input code that you ran to hit this error.

**Your Operating system and DeepLabCut version**

OS: Windows 10
DeepLabCut Version: 2.2b8
Anaconda env used: The one DLC provides (DLC-GPU)

**Describe the problem**

Hello. I have been using DLC for a while now, though I still consider myself a novice. I have been using MA DLC to track groups of vervet monkeys, and have been having a pretty good time with the program so far, apart from a few issues which have been resolved. I have extracted and manually labeled around 200 frames (20 frames or so from each of 10 videos, depending on the video length). A couple of the longer videos I have labeled around 50 frames). I created the training dataset, and then trained the network for 100,000 iterations on my GTX 1070TI. I proceeded to evaluate the network and cross-validate as normal. 

I ran the analyse videos function on one of the shorter videos using both box and skeletal tracking. Both provided similar results. I was also sure to select "create video for checking detections". After analyzing, I watched the generated video to verify that the pose estimation was good. I was incredibly happy with the quality of the markers on the vervets, though there was some jitter, which I expected. Nonetheless, all of the colored markers appeared to be in the correct spots at least 90% of the time. 

Then I proceed to convert to tracklets and attempt to refine. This is where my problem began. There were either very few tracklets, none at all, or they were entirely in the wrong locations when compared to the 'create video for checking detections'. I did some tinkering with the inference_config.yaml, and attempted to follow some advice found in the forums and in the github issues. Nothing I tried seemed to resemble the great tracking I received from 'create video for checking detections'. I am wondering what is going wrong here? Refining tracklets at this point (even with my best variant of the inference_config) will involve moving each and every tracklet, reassigning entities, etc for nearly every frame, not to mention you cannot replace missing tracklets in the Refine Tracklets GUI for ones that are not generated by the system. Is this the intended workflow? Have I made a mistake is training? I am really not certain. I figured 200 labelled frames would be enough (which it seems to be for creating video for checking detections). 

Anyhow, I would really appreciate any input, advice, or help that you are willing to provide. 

Hope to hear back from you.

**Inference_config**
Here is the content of the inference_config.yaml that I am using. Most of it was generated by the system, but I have been tinkering with the values that are not cross validated (like pafthreshold).

```
variant: 0
minimalnumberofconnections: 4
averagescore: 0.1
distnormalization: 1500
distnormalizationLOWER: 0
detectionthresholdsquare: 0
addlikelihoods: 0.15
pafthreshold: 0.03

method: m1
withid: false
topktoretain: 3
upperbound_factor: 1.25
lowerbound_factor: .75
boundingboxslack: 0
max_age: 200
min_hits: 3
iou_threshold: 0.2
```

**config.yaml**
```
# Project definitions (do not edit)
Task: Vigilence
scorer: Calvin
date: Jul2
multianimalproject: true

    # Project path (change when moving around)

project_path: S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02

    # Annotation data set configuration (and individual video cropping parameters)
video_sets:
  ? S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\ITE do not mind
    my comments!_cropped\ITE do not mind my comments!_cropped.mp4
  : crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\20170626 095735_cropped\20170626 095735_cropped.mp4:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\20170922_133351_cropped\20170922_133351_cropped.mp4:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\IMG_2379_cropped\IMG_2379_cropped.MOV:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P6160026_cropped\P6160026_cropped.MOV:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P6270395_cropped\P6270395_cropped.MOV:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P7260647_cropped\P7260647_cropped.MOV:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P9180068_cropped\P9180068_cropped.MOV:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\PA030770_cropped\PA030770_cropped.MOV:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\PA191044_cropped\PA191044_cropped.MOV:
    crop: 0, 400, 0, 400
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\PA231187_cropped\PA231187_cropped.MOV:
    crop: 0, 400, 0, 400
individuals:
- vervet1
- vervet2
- vervet3
- vervet4
- vervet5
- vervet6
- vervet7
- vervet8
- vervet9
- vervet10
- vervet11
- vervet12
- vervet13
- vervet14
- vervet15

uniquebodyparts: []
multianimalbodyparts:
- lefteye
- righteye
- leftear
- rightear
- nose
- chin
- lefthand
- leftelbow
- leftshoulder
- righthand
- rightelbow
- rightshoulder
- topback
- midback
- lowerback
- leftfoot
- leftknee
- lefthip
- rightfoot
- rightknee
- righthip
- tailbase
- tailmid
- tailtip
skeleton:
- - lefteye
  - righteye
- - lefteye
  - nose
- - righteye
  - nose
- - lefteye
  - leftear
- - righteye
  - rightear
- - leftear
  - topback
- - rightear
  - topback
- - nose
  - chin
- - nose
  - leftear
- - nose
  - rightear
- - chin
  - leftshoulder
- - chin
  - rightshoulder
- - chin
  - topback
- - leftear
  - leftshoulder
- - rightear
  - rightshoulder
- - chin
  - leftear
- - chin
  - rightear
- - rightshoulder
  - leftshoulder
- - rightshoulder
  - rightelbow
- - rightelbow
  - righthand
- - rightelbow
  - leftelbow
- - righthand
  - lefthand
- - leftshoulder
  - leftelbow
- - leftshoulder
  - lefthand
- - leftshoulder
  - topback
- - rightshoulder
  - righthand
- - rightshoulder
  - topback
- - topback
  - midback
- - topback
  - righthip
- - topback
  - lefthip
- - lowerback
  - midback
- - lowerback
  - lefthip
- - lowerback
  - righthip
- - lowerback
  - tailmid
- - lowerback
  - tailbase
- - leftelbow
  - lefthand
- - rightear
  - midback
- - leftear
  - midback
- - leftshoulder
  - midback
- - leftshoulder
  - lefthip
- - rightshoulder
  - righthip
- - rightshoulder
  - midback
- - midback
  - tailbase
- - midback
  - lefthip
- - midback
  - righthip
- - lefthip
  - righthip
- - lefthip
  - leftknee
- - lefthip
  - tailbase
- - righthip
  - tailbase
- - righthip
  - rightfoot
- - lefthip
  - leftfoot
- - leftknee
  - leftfoot
- - righthip
  - rightknee
- - rightknee
  - rightfoot
- - rightknee
  - leftknee
- - rightfoot
  - leftfoot
- - rightfoot
  - righthand
- - leftfoot
  - lefthand
- - tailbase
  - tailmid
- - tailmid
  - tailtip
- - tailbase
  - tailtip
- - righthip
  - tailmid
- - righthip
  - tailtip
- - lefthip
  - tailmid
- - lefthip
  - tailtip

bodyparts: MULTI!
start: 0
stop: 1
numframes2pick: 30

    # Plotting configuration
skeleton_color: black
pcutoff: 0.6
dotsize: 3
alphavalue: 0.7
colormap: plasma

    # Training,Evaluation and Analysis configuration
TrainingFraction:
- 0.95
iteration: 0
default_net_type: resnet_50
default_augmenter: multi-animal-imgaug
snapshotindex: -1
batch_size: 8

    # Cropping Parameters (for analysis and outlier frame detection)
cropping: false
croppedtraining: true
    #if cropping is true for analysis, then set the values here:
x1: 0
x2: 640
y1: 277
y2: 624

    # Refinement configuration (parameters from annotation dataset configuration also relevant in this stage)
corner2move2:
- 50
- 50
move2corner: true
video_sets_original:
  ? S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\ITE do not mind
    my comments!\ITE do not mind my comments!.mp4
  : crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\20170626 095735\20170626 095735.mp4:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\20170922_133351\20170922_133351.mp4:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\IMG_2379\IMG_2379.MOV:
    crop: 0, 1280, 0, 720
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P6160026\P6160026.MOV:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P6270395\P6270395.MOV:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P7260647\P7260647.MOV:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\P9180068\P9180068.MOV:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\PA030770\PA030770.MOV:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\PA191044\PA191044.MOV:
    crop: 0, 1920, 0, 1080
  S:\Documents\Barret-HenziRA\Vigilence-Calvin-2020-07-02\videos\PA231187\PA231187.MOV:
    crop: 0, 1920, 0, 1080
```

Not sure if these will help to evaluate the issue or not.

**Comment by jeylau** [comment]
Hello @CalvinLloyd117! Since detections look great, you could give an additional check by running `deeplabcut.utils.make_labeled_video.create_video_from_pickled_tracks` (and passing for example the _box.pickle file); the video takes a bit longer to create, but it will let you evaluate whether the tracker does a good job.
Also, what parameters did you use when refining the tracklets?

**Comment by CalvinLloyd117** [comment]
So I ran the following function

`deeplabcut.utils.make_labeled_video.create_video_from_pickled_tracks('videos/PA231187/PA231187.MOV','videos/PA231187/PA231187DLC_resnet50_VigilenceJul2shuffle1_100000_bx.pickle', pcutoff=0.6)`

Unfortunately, the video that the function produced was unplayable (not sure what is going wrong here). Fortunately, I was still able to inspect the frames that the function was generating with the markers. It was similar to the results from the tracklets GUI. There was a lot of missing data, and several identity swaps given by the color of the markers for a monkey changing frame to frame. Here are some examples of frames produced by the function:

Frame 660:
![frame0660](https://user-images.githubusercontent.com/45299786/94948188-29586400-049c-11eb-8fd7-bf2d439be947.png)

Frame 673:
![frame0673](https://user-images.githubusercontent.com/45299786/94948129-0af26880-049c-11eb-8029-93d2ff04ef00.png)

Frame 740:
![frame0740](https://user-images.githubusercontent.com/45299786/94948236-3e34f780-049c-11eb-910f-39354313e5cc.png)

Here is an image of a frame from PA231187DLC_resnet50_VigilenceJul2shuffle1_100000_full.mp4 (for reference):
![Screenshot (64)](https://user-images.githubusercontent.com/45299786/94948449-94099f80-049c-11eb-9d07-40d70ef943b2.png)
Unfortunately, I don't have access to the actual frame number in this case, but hopefully this can provide some comparison.

To answer the next part of your question, I used the supplied defaults, but did some tinkering with the max gap size to see if there was any improvements. While it did change the output, it never actually "helped"... I Just decided to keep the default values that the GUI provides for refinement.

**Comment by jeylau** [comment]
Alright, so it is an issue with tracking rather than linking tracklets. Would you mind trying with min_hits=1 and a much smaller max_age (e.g. 20)? Also, what is the value of distnorm in the sd.csv file you can find in the folder evaluation-results?

**Comment by CalvinLloyd117** [comment]
Alright, So I ran "convert to tracklets" (I did not re-analyse), ensuring that "Overwrite tracking files" was set to 'Yes'. There appears to be some minor improvements here at first glance, specifically that there is far less missing data, though some is still missing. Because there are so many tracklets, they also appear to be in the wrong places far more often. Additionally, there are more entity swaps, and vervets are being labeled multiple times (ex. A single monkey being labeled as vervet1 and vervet2).

Also, I cannot find the sd.csv file. :(

**Comment by jeylau** [comment]
Just make sure you run `deeplabcut.evaluate_multianimal_crossvalidate`. This should create the csv file and correctly set `distnormalization` in the inference_cfg.yaml. I guess 1500 was manually set, but it may be too small as your images are very big :) I hope that helps!

**Comment by CalvinLloyd117** [comment]
I ran `deeplabcut.evaluate_multianimal_crossvalidate('config.yaml')` and it did not produce the csv you mentioned. The only csv in evaluation-results is 'results.csv'. I took your advice and increased the `distnormalization`. No noticeable improvement unfortunately...  The biggest issue is double labeling (and missing labels). Is there any value that prevents multiple labels from different entities from occupying the same space?

My videos also contain varying amounts of vervets. The max amount there was in a single video was 15, and so I trained the network assuming that max. That is also why my config.yaml contains 15 entities. I am curious if there was a way to set the max amount of entities on a video-to-video basis post-training?  Again, any help you are able to offer is much appreciated. Thank you for your continued patience and help. :)

**Comment by jeylau** [comment]
Aw my bad, the csv file is actually produced after evaluating the network 😑. If you had evaluated it prior to 2.2b8, the file should be found in dlc-models/.../test/.
distnormalizationLOWER is the parameter you are looking for (have a look at the default template for some short descriptions [https://github.com/DeepLabCut/DeepLabCut/blob/master/deeplabcut/inference_cfg.yaml](url)).
As for the ability to change the max number of animals per video, you could edit 'topktoretain' in the inference_cfg (in the dlc-models/.../test folder) and `convert_detections2tracklets` video by video.

**Comment by MMathisLab** [comment]
any updates, @CalvinLloyd117 ?

**Comment by CalvinLloyd117** [comment]
Hello again! Thanks for taking the time to answer some of my questions about the refinement. I took the advice, and the output is looking okay. I also went into the config file, and reduced the number of vervets in there to three (down from 15) and that drastically reduced the amount of multi-labeling I was getting. Unfortunately, I am still having issues with missing labels, labels in the wrong places, and incorrect identities etc, but it is much more manageable. I ended up tinkering with the inference file a bit more as well to really try to improve the tracking.

This is the inference file I decided to go with in the end, in case you were curious:
```
variant: 0
minimalnumberofconnections: 4
averagescore: 0.1
distnormalization: 1500
distnormalizationLOWER: 200
detectionthresholdsquare: 0
addlikelihoods: 0.15
pafthreshold: 0.015
method: m1
withid: false
topktoretain: 3
upperbound_factor: 1.25
lowerbound_factor: .75
boundingboxslack: 0
max_age: 20
min_hits: 1
iou_threshold: 0.4
```
 In the end, I just decided to go ahead with the refinement. I am unsure if I have made a mistake in going ahead with the refinement here, but alas... It is quite annoying at the least to go through frame by frame and manually correct the majority of the labels. I am hoping that after refinement and retraining that my results will be a lot better. This is my first refinement phase after all. It is just extremely time-consuming.

I hoped I could ask a few follow up questions, if it is not too much trouble. Firstly, with the number of missing labels, I was wondering if after merging the data set, and retraining, if those missing labels are going to interfere with the results? In general, would a retain after refinement improve results enough to make future refinements more manageable?

**Comment by MMathisLab** [comment]
it should be fine, i.e. not interfere; in refinement, all labels are always shown, you can delete them or adjust them. Also note your effort in correcting any video can be used - just click merge data after each video you edit, it automatically creates a new training set you can use in the future (sorry for my delayed response!)
